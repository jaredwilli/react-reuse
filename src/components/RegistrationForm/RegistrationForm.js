import React from 'react';
import PropTypes from 'prop-types';
import TextInput from '../TextInput';
import PasswordInput from '../PasswordInput';

/** Registration form with built-in validation. */
class RegistrationForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            user: {
                email: '',
                password: ''
            },
            errors: {},
            submitted: false
        };
    }

    onChange = event => {
        const user = this.state.user;
        user[event.target.name] = event.target.value;
        this.setState({ user });
    }

    // Returns a number from 0 to 100 to represent password quality.
    // For simplicity just returning % of minLength entered
    // Could enhance with checks for number, special char, uniques etc.
    passwordQuality(password) {
        if (!password) {
            return null;
        }
        if (password.length >= this.props.minPasswordLength) {
            return 100;
        }
        const percentOfMinLength = parseInt(password.length / this.props.minPasswordLength * 100, 10);
        return percentOfMinLength;
    }

    validate({ email, password }) {
        const errors = {};
        const { minPasswordLength } = this.props;

        if (!email) {
            errors.email = 'Email is required.';
        }
        if (password.length < minPasswordLength) {
            errors.password = `Password must be ${minPasswordLength} characters long.`;
        }

        this.setState({ errors });

        const formIsValid = Object.getOwnPropertyNames(errors).length === 0;
        return formIsValid;
    }

    onSubmit = () => {
        const { user } = this.state;
        const formIsValid = this.validate(user);

        if (formIsValid) {
            this.props.onSubmit(user);
            this.setState({ submitted: true });
        }
    }

    render() {
        const { errors, submitted } = this.state;
        const { email, password } = this.state.user;

        return (
            submitted
                ? <h2>{this.props.confirmationMessage}</h2>
                : <div>
                    <form>
                        <TextInput
                            htmlId="registration-form-email"
                            name="email"
                            onChange={this.onChange}
                            label="Email"
                            value={email}
                            error={errors.email}
                            required
                        />

                        <PasswordInput
                            htmlId="registration-form-password"
                            name="password"
                            value={password}
                            onChange={this.onChange}
                            quality={this.passwordQuality(password)}
                            maxLength={50}
                            error={errors.password}
                            showVisibilityToggle
                        />

                        <input type="submit" value="Register"
                            onClick={this.onSubmit}
                        />
                    </form>
                </div>
        );
    }
}

RegistrationForm.propTypes = {
    /** Message displayed upon successful submission */
    confirmationMessage: PropTypes.string,

    /** Called when form is submitted */
    onSubmit: PropTypes.func.isRequired,

    /** Minimum password length */
    minPasswordLength: PropTypes.number
};

RegistrationForm.defaultProps = {
    confirmationMessage: 'Thanks for registering',
    minPasswordLength: 8
};

export default RegistrationForm;
