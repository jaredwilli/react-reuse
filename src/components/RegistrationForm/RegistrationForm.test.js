import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import RegistrationForm from './RegistrationForm';

describe('RegistrationForm', () => {
    it('renders the component correctly', () => {
        const tree = renderer.create(<RegistrationForm />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
