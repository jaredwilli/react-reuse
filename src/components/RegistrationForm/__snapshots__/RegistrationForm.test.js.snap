// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm renders the component correctly 1`] = `
<div>
  <form>
    <div
      style={
        Object {
          "marginBottom": 16,
        }
      }
    >
      <label
        htmlFor="registration-form-email"
        style={
          Object {
            "display": "block",
          }
        }
      >
        Email
         
        <span
          style={
            Object {
              "color": "red",
            }
          }
        >
           *
        </span>
      </label>
      <input
        id="registration-form-email"
        name="email"
        onChange={[Function]}
        placeholder={undefined}
        style={undefined}
        type="text"
        value=""
      />
    </div>
    <div
      style={
        Object {
          "marginBottom": 16,
        }
      }
    >
      <label
        htmlFor="registration-form-password"
        style={
          Object {
            "display": "block",
          }
        }
      >
        Password
         
        <span
          style={
            Object {
              "color": "red",
            }
          }
        >
           *
        </span>
      </label>
      <input
        id="registration-form-password"
        maxLength={50}
        name="password"
        onChange={[Function]}
        placeholder={undefined}
        style={undefined}
        type="password"
        value=""
      />
      <a
        onClick={[Function]}
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        <svg
          height="16"
          viewBox="0 0 22 22"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            fill="#4d4d4d"
            transform="matrix(.02146 0 0 .02146 1 1)"
          >
            <path
              d="m466.07 161.53c-205.6 0-382.8 121.2-464.2 296.1-2.5 5.3-2.5 11.5 0 16.9 81.4 174.9 258.6 296.1 464.2 296.1 205.6 0 382.8-121.2 464.2-296.1 2.5-5.3 2.5-11.5 0-16.9-81.4-174.9-258.6-296.1-464.2-296.1m0 514.7c-116.1 0-210.1-94.1-210.1-210.1 0-116.1 94.1-210.1 210.1-210.1 116.1 0 210.1 94.1 210.1 210.1 0 116-94.1 210.1-210.1 210.1"
            />
            <circle
              cx="466.08"
              cy="466.02"
              r="134.5"
            />
          </g>
        </svg>
      </a>
    </div>
    <input
      onClick={[Function]}
      type="submit"
      value="Register"
    />
  </form>
</div>
`;
